driver-core/src/main/com/mongodb/connection/BsonWriterHelper.java

```diff
@@ -0,0 +1,113 @@
+
+    BsonWriterHelper(final BsonBinaryWriter writer) {
+        this(writer, Integer.MAX_VALUE);
+    }
+
+    BsonWriterHelper(final BsonBinaryWriter writer, final int maxBatchLength) {
+        this.writer = writer;
+        this.maxBatchLength = maxBatchLength;
+    }
+
+    void writePayloadArray(final SplittablePayload payload, final int commandStartPosition) {
+        writer.writeStartArray(payload.getPayloadName());
+        writePayload(payload, commandStartPosition);
+        writer.writeEndArray();
+    }
+
+    void writePayload(final SplittablePayload payload) {
+        for (int i = 0; i < payload.getPayload().size(); i++) {
+            if (writeDocument(payload.getPayload().get(i), writer.getBsonOutput().getPosition(), i + 1)) {
+                payload.setPosition(i + 1);
+            } else {
+                break;
+            }
+        }
+    }
+
+    void writePayload(final SplittablePayload payload, final int startPosition) {
+        for (int i = 0; i < payload.getPayload().size(); i++) {
+            if (writeDocument(payload.getPayload().get(i), startPosition, i + 1)) {
+                payload.setPosition(i + 1);
+            } else {
+                break;
+            }
+        }
+    }
+
+    boolean writeDocument(final BsonDocument document, final int startPosition, final int batchItemCount) {
+        int currentPosition = writer.getBsonOutput().getPosition();
+        getCodec(document).encode(writer, document, ENCODER_CONTEXT);
+        if (exceedsLimits(writer.getBsonOutput().getPosition() - startPosition, batchItemCount)) {
+            writer.getBsonOutput().truncateToPosition(currentPosition);
+            return false;
+        }
+        return true;
+    }
+
+    void writeElements(final List<BsonElement> bsonElements) {
+        for (BsonElement bsonElement : bsonElements) {
+            writer.writeName(bsonElement.getName());
+            getCodec(bsonElement.getValue()).encode(writer, bsonElement.getValue(), ENCODER_CONTEXT);
+        }
+    }
+
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    Codec<BsonValue> getCodec(final BsonValue bsonValue) {
+        return (Codec<BsonValue>) REGISTRY.get(bsonValue.getClass());
+    }
+
+
+    private boolean exceedsLimits(final int batchLength, final int batchItemCount) {
+        return (exceedsBatchLengthLimit(batchLength, batchItemCount) || exceedsBatchItemCountLimit(batchItemCount));
+    }
+
+    // make a special exception for a command with only a single item added to it.  It's allowed to exceed maximum document size so that
+    // it's possible to, say, send a replacement document that is itself 16MB, which would push the size of the containing command
+    // document to be greater than the maximum document size.
+    private boolean exceedsBatchLengthLimit(final int batchLength, final int batchItemCount) {
+        return batchLength > writer.getBinaryWriterSettings().getMaxDocumentSize() && batchItemCount > 1;
```

Comment:
This looks incorrect, but I'm not sure... The limits the driver has to enforce are:

- the total length of any message can't exceed isMasterResponse.maxMessageSizeBytes
- the total count of writes can't exceed isMasterResponse.maxWriteBatchSize
- the total length of the command document can't exceed isMasterResponse.maxBsonObjectSize
- For updates and deletes, the length of each document in the payload can't exceed isMasterResponse.maxBsonObjectSize + HEADROOM
- For inserts, the length of each document in the payload can't exceed isMasterResponse.maxBsonObjectSize

I'm not sure all of those are being enforced properly.  It looks here, for instance, like it's limiting the total message length to maxBsonObjectSize instead of maxMessageSizeBytes.